controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers"
+	provutils "github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers/utils"
+	rc "github.com/RedHatInsights/rhc-osdk-utils/resource_cache"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	core "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type ClowdAppReconciliation struct {
+	cache                 *rc.ObjectCache
+	recorder              record.EventRecorder
+	ctx                   *context.Context
+	client                client.Client
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
+	cache *rc.ObjectCache,
+	client client.Client,
+	recorder record.EventRecorder,
+	app *crd.ClowdApp,
+	log logr.Logger,
+	req *ctrl.Request) ClowdAppReconciliation {
+	return ClowdAppReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      &ctx,
+		client:   client,
+		log:      &log,
+		app:      app,
+		req:      req,
+	}
+}
+
+func (r *ClowdAppReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.getApp,
+		r.setPresentAndManagedApps,
+		r.startMetrics,
+		r.isAppMarkedForDeletion,
+		r.addFinalizer,
+		r.isEnvLocked,
+		r.isAppDisabled,
+		r.getClowdEnv,
+		r.isEnvCoreNamespaceDeleted,
+		r.isAppCoreNamespaceDeleted,
+		r.isClowdEnvReconciled,
+		r.isEnvReady,
+		r.runProviders,
+		r.applyCache,
+		r.setAppResourceStatus,
+		r.deletedUnusedResources,
+		r.setReconciliationSuccessful,
+		r.stopMetrics,
+	}
+}
+
+func (r *ClowdAppReconciliation) Reconcile() (ctrl.Result, error) {
+	for _, step := range r.steps() {
+		result, err := step()
+		if err != nil {
+			return result, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) startMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics = ReconciliationMetrics{}
+	r.reconciliationMetrics.init(r.app.Name, r.app.Spec.EnvName)
+	r.reconciliationMetrics.start()
+
+	if clowderconfig.LoadedConfig.Features.PerProviderMetrics {
+		requestMetrics.With(prometheus.Labels{"type": "app", "name": r.app.GetIdent()}).Inc()
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) stopMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics.stop()
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) setPresentAndManagedApps() (ctrl.Result, error) {
+	if _, ok := presentApps[r.app.GetIdent()]; !ok {
+		presentApps[r.app.GetIdent()] = true
+	}
+	presentAppsMetric.Set(float64(len(presentApps)))
+
+	delete(managedApps, r.app.GetIdent())
```

Comment:
Can you add a comment here - it took me a moment to grok what I did - we remove it from the managed list because it _may_ have been managed before, but it may not be after this reconcile.

Reply:
Done

Reply:
I don't see a comment :)